project-chip/connectedhomeip
# Add and Enable IM subscription support

#### Problem
Follow IM spec chapter 5 and IM encoding spec 7.3 and 7.4 to achieve
subscription functionality.
https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Interaction-Model.adoc#5-subscribe-interaction
https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Encoding-Specification.adoc#subscriberequestmessage
 #7978  #7979
#### Change overview
- Establishing distinct subscriptions and tracking their respective configurations from multiple clients simultaneously.
- Sharing of reporting logic between reads and subscriptions to maximize re-use.
- Multiple paths per subscription
- Automatic liveness assessment on client and server and termination of subscription accordingly.
- Eventing support in addition to attribute reporting.
- Spec compliance to the encoding specification with TLV-based payloads.
- Preservation of the application level APIs for client-side reporting, but with a swap of the internals to use the new machinery
- Update templates to use interaction model APIs for sending subscribe
requests.
- Update DeviceControllerInteractionModelDelegate to handle report
messages from subscription.
- Update ChipCallback Mgr to bridge subscribe responses to existing
callbacks, and handle TLV message just as read responses.

#### Testing
- Update TestSuite for basic Subscription protocol tests, test routing
looks like follows:
Send subscribe request, wait for response.
Wait for a few seconds, and execute "kick" commands.
The test is expected to receive the same number of Reports as the
  number of kick commands.
The test will fail when not receiving enough report data from server
  in global timeout.
- Update python script test to (in cirque) include a similar test
routine as test suite.
- Add unit test for positive and negative subscribe test
- Add integrated cirque test for subscribe
